### DISCUSSION TITLE
ValueError

### DISCUSSION THREAD

**Discussion by CaiYitao** [cjdoris/PythonCall.jl#143] [Q&A] [answered]
Got the following error, what could be the problem?
the functions are from rdkit
```
@py begin
	import rdkit
	import rdkit.Chem.AllChem as AC
	import rdkit.Chem as Chem
	import numpy as np
	end
```

<img width="717" alt="image" src="https://user-images.githubusercontent.com/65641330/162580372-3a9364ab-7f22-4101-8643-c8c1b12cf3b5.png">

or use julia array:

<img width="655" alt="image" src="https://user-images.githubusercontent.com/65641330/162580391-0bd92b06-58e8-40b6-8369-a606f7fc999c.png">

Thank you!

**Comment by icweaver** [comment]
I think this usually happens when trying to do a py->jl conversion that is not supported. For example:

```julia
julia> using PythonCall

julia> Float64[pyfloat(4)]
ERROR: MethodError: Cannot `convert` an object of type Py to an object of type Float64
Closest candidates are:
  convert(::Type{T}, ::T) where T<:Number at ~/julia-1.7.2/share/julia/base/number.jl:6
  convert(::Type{T}, ::Number) where T<:Number at ~/julia-1.7.2/share/julia/base/number.jl:7
  convert(::Type{T}, ::Base.TwicePrecision) where T<:Number at ~/julia-1.7.2/share/julia/base/twiceprecision.jl:262
  ...
Stacktrace:
 [1] setindex!(A::Vector{Float64}, x::Py, i1::Int64)
   @ Base ./array.jl:903
 [2] getindex(#unused#::Type{Float64}, x::Py)
   @ Base ./array.jl:412
 [3] top-level scope
   @ REPL[2]:1
```

So I think a Julia object might be sneaking into your setup somewhere before it is passed to a python function that does not know how to handle it.

> or use julia array

Same thing here. `rdkit` seems to be expecting a "Numeric array" (is this their lingo for a numpy array?), so passing the Julia array `arr = zeros(1)` instead of the numpy array `arr = np.zeros(1)` will cause `ConvertToNumpyArray` to rightly complain

**Reply by icweaver** [reply]
fwiw, trying a version of the example you shared seems to work alright for me at least:

![screenshot](https://user-images.githubusercontent.com/25312320/162584846-0cdf1e4a-60bc-4095-a41f-706ee6635dfc.png)

<details>
  <summary>Notebook file</summary>
  
 ```julia
### A Pluto.jl notebook ###
# v0.19.0

using Markdown
using InteractiveUtils

# ╔═╡ eb585990-b821-11ec-26ae-e928d6ac0b1d
begin
	using Pkg
	Pkg.activate(Base.current_project())
	
	using PythonCall
end

# ╔═╡ 0e7d0578-1e8f-4864-8047-50f3e8fb6952
@py begin
	import rdkit
	import rdkit.Chem.AllChem as AC
	import rdkit.Chem as Chem
	import numpy as np
end

# ╔═╡ bf1e5c85-3648-456e-8327-343a7c9df595
mol = Chem.MolFromSmiles("C1CCCOC1")

# ╔═╡ 51d28116-5823-4dad-a26c-8b727dba12f1
fp_vec = AC.GetMorganFingerprintAsBitVect(mol, radius=3, nBits=8)

# ╔═╡ 66288740-497b-454c-a6d9-b8c5e92141a8
begin
	arr = np.zeros(1)
	# I guess this mutates
	Chem.DataStructs.ConvertToNumpyArray(fp_vec, arr)
	arr
end

# ╔═╡ 51991470-a8fd-4e10-ae5a-e7c0a0b06ccf
print(fp_vec)

# ╔═╡ Cell order:
# ╠═eb585990-b821-11ec-26ae-e928d6ac0b1d
# ╠═0e7d0578-1e8f-4864-8047-50f3e8fb6952
# ╠═bf1e5c85-3648-456e-8327-343a7c9df595
# ╠═51d28116-5823-4dad-a26c-8b727dba12f1
# ╠═66288740-497b-454c-a6d9-b8c5e92141a8
# ╠═51991470-a8fd-4e10-ae5a-e7c0a0b06ccf
```
</details>

I don't know anything about chem, or what global variables are being passed in your screenshot, so I could be misunderstanding what is causing your issue here

**Comment by CaiYitao** [comment]
This is amazing! Thank you very much for your help!
Now I found that when we use for loop it spits that error, if we do not use for loop, it just works as you showed.

**Comment by cjdoris** [comment, marked as answer]
Your first error was on line 8, the one below the one highlighted, namely `fps[i] = arr`, which is complaining that you're trying to set an element of an array of floats to a Python object.

**Reply by CaiYitao** [reply]
Yeah, Thank you for pointing out, yeah, the a was mutated to an array of shapes (1, L), now this is interesting Python can do this
<img width="751" alt="image" src="https://user-images.githubusercontent.com/65641330/162589037-35094747-6d45-4cd0-9c46-8179e732e578.png">
<img width="559" alt="image" src="https://user-images.githubusercontent.com/65641330/162589326-d597e6bc-4847-4dd9-a720-9c05713147b6.png">


After I convert the python array into julia type Array, it gave this error
<img width="712" alt="image" src="https://user-images.githubusercontent.com/65641330/162589133-bbe2cb26-207f-4ab1-9fb1-59d6b5c8f916.png">

but after I change the 
```begin
	Lₚ = 1024
	fps = np.zeros((length(y), Lₚ))
end
```
<img width="642" alt="image" src="https://user-images.githubusercontent.com/65641330/162589550-5ddcedea-9b93-4ff4-864a-1acb3df5b42b.png">
it works fine

Really appreciate your help!!!

**Reply by icweaver** [reply]
> Your first error was on line 8, the one below the one highlighted, namely fps[i] = arr

oop, missed that line when transcribing the screenshot. Thanks @cjdoris!
